include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+     */
+    constexpr inline bool operator==(const sidb_cluster_charge_state& other) const noexcept
+    {
+        return static_cast<uint64_t>(*this) == static_cast<uint64_t>(other);
+    }
+    /**
+     * Defined a hashing of a cluster charge state. Since we need only separate cluster charge states by their
+     * compressed form, we may compute a hash over this for optimal performance when used in a hash set.
+     *
+     * @param m Cluster charge state to compute the hash of.
+     * @return The hash of the given cluster charge state.
+     */
+    inline std::size_t operator()(const sidb_cluster_charge_state& m) const noexcept
+    {
+        return std::hash<uint64_t>{}(static_cast<uint64_t>(m));
+    }
+    /**
+     * Defines addition of cluster charge states through multiset concatenation. Disregards compositions.
+     *
+     * @param other Other cluster charge state to concatenate with the current.
+     * @return The concatenated cluster charge state, which is the modified version of the current.
+     */
+    constexpr inline sidb_cluster_charge_state& operator+=(const sidb_cluster_charge_state& other) noexcept
+    {
+        neg_count += other.neg_count;
+        pos_count += other.pos_count;
+        return *this;
+    }
+    /**
+     * Defines subtraction of cluster charge states through multiset difference. Disregards compositions.
+     *
+     * @param other Other cluster charge state to take the difference of w.r.t. with the current.
+     * @return The cluster charge state that is their difference, which is the modified version of the current.
+     */
+    constexpr inline sidb_cluster_charge_state& operator-=(const sidb_cluster_charge_state& other) noexcept
+    {
+        assert(neg_count >= other.neg_count && pos_count >= other.pos_count);
+        neg_count -= other.neg_count;
+        pos_count -= other.pos_count;
+        return *this;
+    }
+};
+/**
+ * Function to convert a singleton cluster charge state in its compressed form to a charge state.
+ *
+ * @param m A singleton multiset charge configuration.
+ * @return The charge state associated with the sole element contained in the given multiset charge configuration.
+ */
+static constexpr inline sidb_charge_state singleton_multiset_conf_to_charge_state(const uint64_t m) noexcept
+{
+    return sign_to_charge_state(
+        static_cast<int8_t>(static_cast<uint32_t>(m) - static_cast<uint32_t>(static_cast<uint32_t>(m) < m)));
+}
+/**
+ * This struct defines the type of a potential projection, which pairs a multiset charge configuration with the
+ * potential value (in eV) associated with the potential projection in the given context. The context is given by
+ * potential projection stores in the decorated cluster hierarchies, which links the projecting cluster with SiDB that
+ * receives this potential projection.
+ */
+struct potential_projection
+{
+    /**
+     * Potential projection value (unit: eV).
+     */
+    double v{0.0};
```

Comment:
Can we use something other than ``v``?

Reply:
`pot_val` is what I put in the upcoming push. Along with `multiset` instead of `m`